### PR TITLE
refactor: route per-LLMQ-type caches through PerLlmqTypeCache

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented

**Note on scope after rebase:** this PR originally carried the QSIGSHARE fix itself. That fix shipped independently in v23.1.8 (`a801d3f3b72`) and reached `develop` through the master back-merge in #7534, so the handler gate and the two `find()`-guarded lookups are already present. What remains here is the structural follow-up, retitled accordingly.

`Consensus::LLMQType` is a `uint8_t` enum serialized verbatim over the wire with no range check. The LLMQ LRU caches keyed by it are `std::map<LLMQType, unordered_lru_cache<...>>`, seeded only with the types registered for the active chain. `operator[]` on an unregistered key default-constructs the cache with `MaxSize = 0`, and `unordered_lru_cache`'s constructor runs `assert(_maxSize != 0)`. Since `src/util/check.h` makes compiling with `NDEBUG` a hard `#error`, that assert is live in release builds — so any such `operator[]` is a deterministic abort waiting for a caller that has not validated the type.

The v23.1.8 fix closed the reported vector by gating the QSIGSHARE handler and switching the two lookups a `QSIGSHARE` could reach (`HasMinedCommitment`, `GetQuorum`/`GetCachedMutableQuorum`) to `find()`. The remaining caches still use `operator[]`, and nothing in the type system stops a future caller from reaching one with an unvalidated type.

## What was done?

Remove the hazard structurally rather than gating each lookup. A new `PerLlmqTypeCache<Value, Key>` (`src/llmq/cache.h`) owns the map, holds one LRU per registered type, and answers for any other type as a miss with writes dropped — so no caller can size a zero-capacity cache into existence.

All eight caches are converted (`mapHasMinedCommitmentCache`, `m_qc_hashes_lru`, `mapQuorumsCache`, `scanQuorumsCache`, `mapQuorumMembers`, `mapIndexedQuorumMembers`, `indexed_quorums_cache`, `cleanupQuorumsCache`), the ad-hoc `find()` guards added in v23.1.8 are folded into it, and `InitQuorumsCache` is deleted.

Per-type cache capacities, lock scopes, `GUARDED_BY` annotations and the lazy `if (empty()) Init(...)` seeding are preserved at every converted site.

One deliberate behaviour change worth calling out: `HasMinedCommitment` on an unregistered type now falls through to the EvoDB probe and returns false, where `develop` currently returns false before touching the DB. That probe is the same one any cache miss performs, and the handlers reject unregistered types before reaching it, so this is not remotely reachable — but it is a step back from the early return and is easy to restore if reviewers prefer.

`llmq_invalid_type_tests` is extended to pin the `PerLlmqTypeCache` contract itself (unregistered types read as misses, writes dropped, capacity 0) alongside the existing `HasMinedCommitment` / `GetQuorum` / `GetCachedMutableQuorum` coverage. An earlier revision also carried a unit test for the QSIGSHARE handler ban with its own peer/stream scaffolding; it was dropped per review feedback — comprehensive `NetSigning` regression coverage is better done as its own effort than by pinning one corner case here.

## How Has This Been Tested?

- Full local build clean on macOS/arm64.
- `src/test/test_dash --run_test=llmq_invalid_type_tests,llmq_utils_tests`: 18 test cases, no errors — also exercised by the CI unit-test jobs (linux64, tsan, sqlite, multiprocess), all green on this tree.
- The current head is a commit-message-only amend of the CI-tested commit (identical tree).

## Breaking Changes

None. Only lookups naming an LLMQ type that is not registered for the active chain are affected, and those were never valid.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit eec5fe06)

File: src/llmq/net_signing.cpp [hunk 1]
```diff
@@ -76,6 +77,15 @@ void NetSigning::ProcessMessage(CNode& pfrom, const std::string& msg_type, CData
...
 
         for (const auto& sigShare : receivedSigShares) {
+            // Gate unregistered types like the sibling handlers (QSIGREC, QSIGSESANN,
+            // QGETDATA, QFCOMMITMENT) do, so a wire-supplied type never reaches the quorum
+            // lookups.
+            if (!Params().GetLLMQ(sigShare.getLlmqType()).has_value()) {
+                LogPrint(BCLog::LLMQ_SIGS, "NetSigning::%s -- rejected %s from peer=%d: invalid llmqType [%d]\n",
+                         __func__, msg_type, pfrom.GetId(), std23::to_underlying(sigShare.getLlmqType()));
+                BanNode(pfrom.GetId());
+                return;
```
Comment (thepastaclaw) [inline src/llmq/net_signing.cpp:89]: **🟡 Suggestion: Add coverage for the QSIGSHARE validation and ban path itself**

The new unit suite tests unsupported types only through `PerLlmqTypeCache`, `HasMinedCommitment`, `GetQuorum`, and `GetCachedMutableQuorum`; it never sends an invalid QSIGSHARE through `NetSigning::ProcessMessage`. The existing functional QSIGSHARE case uses a registered type. This leaves the new trust-boundary behavior unprotected: if this gate were removed, the hardened cache would safely miss, `ProcessMessageSigShare` would return true when `GetQuorum` returns null (`signing_shares.cpp` lines 437-440), and the sender would be silently dropped without reaching the fallback `BanNode` call. Add a functional P2P case with SPORK_21 active that sends a `CSigShare` carrying an unregistered type and verifies the rejection log and 100-point peer score; the whitelisted-peer and deprecated-banscore pattern in `feature_llmq_dkg_intake.py` keeps that score observable.

<sub>source: ['claude', 'codex']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex + Sonnet

Carried forward from the prior review, the network-facing invalid-QSIGSHARE rejection and 100-point peer-scoring branch remains untested; no correctness blocker was found in the implementation. In the latest delta, squashing resolves the earlier corrective-follow-up history issue but newly removes the two-commit separation that the PR description explicitly identifies as necessary for minimal backports.
Source: reviewers codex/gpt-5.6-sol (general and dash-core-commit-history) and claude/claude-sonnet-5 (general and dash-core-commit-history; the first specialist attempt was unparseable and a fresh retry succeeded); verifier codex/gpt-5.6-sol. The openclaw-agent coordinator is orchestration-only.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — final-verifier (fallback)
- Sonnet reviewers: `claude-sonnet-5` — general (completed), `claude-sonnet-5` — dash-core-commit-history (failed), `claude-sonnet-5` — dash-core-commit-history (completed)

🟡 2 suggestion(s)

_1 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/llmq/net_signing.cpp`:
- [SUGGESTION] src/llmq/net_signing.cpp:83-87: Add coverage for the QSIGSHARE validation and ban path itself
  The new unit suite tests unsupported types only through `PerLlmqTypeCache`, `HasMinedCommitment`, `GetQuorum`, and `GetCachedMutableQuorum`; it never sends an invalid QSIGSHARE through `NetSigning::ProcessMessage`. The existing functional QSIGSHARE case uses a registered type. This leaves the new trust-boundary behavior unprotected: if this gate were removed, the hardened cache would safely miss, `ProcessMessageSigShare` would return true when `GetQuorum` returns null (`signing_shares.cpp` lines 437-440), and the sender would be silently dropped without reaching the fallback `BanNode` call. Add a functional P2P case with SPORK_21 active that sends a `CSigShare` carrying an unregistered type and verifies the rejection log and 100-point peer score; the whitelisted-peer and deprecated-banscore pattern in `feature_llmq_dkg_intake.py` keeps that score observable.

In `<commit:d8f8044>`:
- [SUGGESTION] <commit:d8f8044>:1: Single squashed commit undermines the stated backport-separability goal
  The PR description says this is structured as two commits: a standalone QSIGSHARE gate small enough to backport, followed by the broader cache refactor and its tests. Current head instead contains exactly one commit whose patch combines the handler gate with the new cache abstraction, eight cache migrations, build wiring, and unit tests. That removes the independently cherry-pickable security fix and conflicts with the stated backport plan. Split the history into a minimal gate commit and a separate `PerLlmqTypeCache` refactor-and-tests commit; this also keeps the broad structural hardening independently reviewable and bisectable.
```
</details>

**Comment by PastaPastaPasta** [comment]
Good catch, added in c7222219f311.

Agreed that the gate was unpinned. I confirmed the mechanism you describe: `CSigSharesManager::ProcessMessageSigShare` returns `true` when `GetQuorum` yields nullptr (`signing_shares.cpp:437-439`), so with the gate removed the sender is dropped silently and never reaches `BanNode`.

I implemented it as a unit case rather than a functional one. `qsigshare_unknown_llmq_type_bans_peer` in `src/test/llmq_invalid_type_tests.cpp` installs a `NetSigning` over `RegTestingSetup`, activates `SPORK_21_QUORUM_ALL_CONNECTED`, feeds a `CSigShare` carrying an unregistered type through `ProcessMessage`, and asserts the peer's misbehaviour score goes 0 -> 100. That covers the same trust boundary without adding a node-spinning functional case; the pattern follows `llmq_qgetdata_tests.cpp` from #7519.

Verified it actually pins the behaviour rather than just passing: with the gate deleted the test fails with `check stats.m_misbehavior_score == 100 has failed [0 != 100]`, and passes again once restored.

One clarification on scope, since the phrasing suggests otherwise: what was untested is the **ban**, not the crash. The `operator[]` -> `find()` hardening in `HasMinedCommitment` / `GetQuorum` / `GetCachedMutableQuorum` is what prevents the `assert(_maxSize != 0)` abort, and that is covered by the existing cases here. Removing the gate would have degraded "ban the peer" to "silently ignore it" -- worth fixing, but not a return of the remote-abort vector.

---
🤖 Posted autonomously by Claude on behalf of pasta.

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Codex + Sonnet

This PR replaces eight ad-hoc per-LLMQ-type caches with a single PerLlmqTypeCache abstraction in one coherent commit (40d714a8778), preserving capacities, lock scopes, and lazy-init semantics at every converted call site, and adds a genuine end-to-end test (qsigshare_unknown_llmq_type_bans_peer) that drives NetSigning::ProcessMessage with a hand-serialized unregistered-type QSIGSHARE message and asserts the peer misbehavior score moves 0→100. Both fresh Sonnet reviewer lanes and the fresh Codex checkpoint independently confirm zero new findings at head 40d714a8778023aaa293ce4d05fa962e76a461ce, and I independently verified: (1) the new test at src/test/llmq_invalid_type_tests.cpp:170-201 pins the handler gate and BanNode's 100-point scoring, and (2) `git log --oneline fff866495dc..40d714a8778 -- src/llmq/net_signing.cpp` returns empty while `git merge-base --is-ancestor a801d3f3b72 fff866495dc` confirms the QSIGSHARE point fix already shipped via v23.1.8/#7534, so this branch carries only the structural refactor with no point fix left to separate. No blocking, suggestion, or nitpick findings apply to the current head.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `claude-sonnet-5` — final-verifier
- Sonnet reviewers: `claude-sonnet-5` — general (completed), `claude-sonnet-5` — dash-core-commit-history (completed)

File: src/llmq/net_signing.cpp (hunk 1 above)
Comment (thepastaclaw) [reply src/llmq/net_signing.cpp:89]: Resolved in this update — _Add coverage for the QSIGSHARE validation and ban path itself_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

File: src/test/llmq_invalid_type_tests.cpp [hunk 2]
```diff
@@ -4,83 +4,201 @@
...
+
+/**
+ * The handler gate itself, not just the caches behind it.
+ *
+ * The cache hardening above makes an unregistered type a safe miss, but a miss alone is not the
+ * contract: CSigSharesManager::ProcessMessageSigShare returns true when GetQuorum yields nullptr,
+ * so without the gate the sender is dropped silently and never scored. Deleting the gate would
+ * leave every other test in this file green, so pin the ban here.
+ */
+BOOST_FIXTURE_TEST_CASE(qsigshare_unknown_llmq_type_bans_peer, RegTestingSetup)
```
Comment (knst) [inline src/test/llmq_invalid_type_tests.cpp]: why changes in the cache of llmq quorums required new regression tests that invoving `net_signing`,  and `peermanager` ?

Comment (PastaPastaPasta) [reply src/test/llmq_invalid_type_tests.cpp]: see pr desc:

> Note on scope after rebase: this PR originally carried the QSIGSHARE fix itself. That fix shipped independently in v23.1.8 (a801d3f3b72) and reached develop through the master back-merge in https://github.com/dashpay/dash/pull/7534, so the handler gate and the two find()-guarded lookups are already present. What remains here is the structural follow-up, retitled accordingly.

I believe that these tests are related to that fix, but was not included in the v23.1.8 commit, etc. so some things are followup to that fix, most is refactor

File: src/test/llmq_invalid_type_tests.cpp [hunk 3]
```diff
@@ -4,83 +4,201 @@
...
 
 BOOST_AUTO_TEST_SUITE(llmq_invalid_type_tests)
 
-// Not a named enumerator, and never present in any chain's consensus params.
+// Unassigned enum values; each test re-asserts they are unregistered, so this stays honest
+// if a future chainparams change starts using one.
 static constexpr Consensus::LLMQType UNKNOWN_LLMQ_TYPE{static_cast<Consensus::LLMQType>(0)};
-// LLMQ_25_67. A real enumerator, but registered on testnet only, so it is unknown under regtest.
-static constexpr Consensus::LLMQType UNREGISTERED_LLMQ_TYPE{Consensus::LLMQType::LLMQ_25_67};
+static constexpr Consensus::LLMQType UNKNOWN_LLMQ_TYPE_ALT{static_cast<Consensus::LLMQType>(55)};
```
Comment (knst) [inline src/test/llmq_invalid_type_tests.cpp]: why unregistered type here is replaced by unknown type? Is it breaking change?

File: src/test/llmq_invalid_type_tests.cpp (hunk 2 above)
Comment (knst) [reply src/test/llmq_invalid_type_tests.cpp]: okie, all of that code is added for these only lines:

```
    BOOST_REQUIRE(m_node.peerman->GetNodeStateStats(peer->GetId(), stats));
    BOOST_CHECK_EQUAL(stats.m_misbehavior_score, 0);

    auto stream = MakeQSigShareStream(UNKNOWN_LLMQ_TYPE);
    net_signing.ProcessMessage(*peer, NetMsgType::QSIGSHARE, stream); <------ test itself

    BOOST_REQUIRE(m_node.peerman->GetNodeStateStats(peer->GetId(), stats));
    BOOST_CHECK_EQUAL(stats.m_misbehavior_score, 100);
```

which actually test these lines:
```
            if (!Params().GetLLMQ(sigShare.getLlmqType()).has_value()) {         
                LogPrint(BCLog::LLMQ_SIGS, "NetSigning::%s -- invalid llmqType[%d] from peer=%d\n", __func__,
                         std23::to_underlying(sigShare.getLlmqType()), pfrom.GetId());
                BanNode(pfrom.GetId());                                          
                return;
            }
```


I think the proper solution is create regressions tests for NetSigning component [not in this PR as out-of-scope] rather than regression test just one corner case.
See:
```
$ grep -r BanNode src/llmq/net_signing.cpp  -c
12
```

I'd recomment to remove MakeSecretKey, MakeQSigShareStream, MakePeer helpers, related headers and qsigshare_unknown_llmq_type_bans_peer test suit and consider creating comprehensive tests for NetSigning ; or maybe cover it in functional tests.

**Review by knst — COMMENTED** [review]
LGTM overall, but check https://github.com/dashpay/dash/pull/7516/files#r3728239626 and https://github.com/dashpay/dash/pull/7516/files#r3723835253

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

The PerLlmqTypeCache refactor raises no validated correctness or security issues at the exact reviewed head. The sole proposed documentation nit is out of scope because the same comment was already inaccurate at the merge base, where the QSIGSHARE-reachable lookups were protected by find()-guarded misses.
Source: reviewers codex/gpt-5.6-sol (general and dash-core-commit-history); final verifier codex/gpt-5.6-sol. openclaw-agent/cliproxy/gpt-5.6-sol is orchestration-only and not reviewer evidence.

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)

**Review by knst — APPROVED** [review]
utACK acd47695276bcdd4999ad2a0796e32ae165b01fe

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

The exact-head refactor consistently routes all eight targeted caches through PerLlmqTypeCache while preserving capacities, locking, lazy initialization, and cache-clearing behavior. Unregistered LLMQ types safely produce misses and dropped writes; the changed objects compile, the targeted 18 unit-test cases pass, and no in-scope defects were identified.
Source: reviewers codex/gpt-5.6-sol (general and dash-core-commit-history); final verifier claude/claude-sonnet-5. openclaw-agent/cliproxy/gpt-5.6-sol is orchestration-only and not reviewer evidence.

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)

**Comment by PastaPastaPasta** [comment]
merged via d1f7ff56d0491af2028f980e1c8763618d71372f; something weird happened where PR didn't close.

**Comment by UdjinM6** [comment]
> merged via [d1f7ff5](https://github.com/dashpay/dash/commit/d1f7ff56d0491af2028f980e1c8763618d71372f); something weird happened where PR didn't close.

I guess it's because you merged the wrong commit, a pre-force-pushed one, see:

> PastaPastaPasta force-pushed the sec/v005 branch from acd4769 to eec5fe0 

https://github.com/dashpay/dash/pull/7516#event-29992207696

and

> 2 parents 6d712a2 + acd4769 commit d1f7ff5

https://github.com/dashpay/dash/commit/d1f7ff56d0491af2028f980e1c8763618d71372f

**Comment by PastaPastaPasta** [comment]
Yes.. I saw that, but not sure why. weird.
